File: 2025-07/0705/0705.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim();

function solve(input) {
  const lines = input.split('\n');
  const N = parseInt(lines[0]);
  const board = lines.slice(1).map((line) => line.split(' ').map(Number));

  const visited = Array.from({ length: N }, () => Array(N).fill(false));
  const queue = [[0, 0]];
  visited[0][0] = true;

  while (queue.length > 0) {
    const [x, y] = queue.shift();
    const jump = board[x][y];

    if (jump === -1) {
      return true; // HaruHaru
    }

    const dx = [0, 1]; // 아래, 오른쪽
    const dy = [1, 0];

    for (let i = 0; i < 2; i++) {
      const nx = x + dx[i] * jump;
      const ny = y + dy[i] * jump;

      if (nx >= 0 && nx < N && ny >= 0 && ny < N && !visited[nx][ny]) {
        visited[nx][ny] = true;
        queue.push([nx, ny]);
      }
    }
  }

  return false; // Hing
}

if (solve(input)) {
  console.log('HaruHaru');
} else {
  console.log('Hing');
}

File: 2025-07/07-05/0705.py
from collections import deque

def bfs():
    q = deque()
    q.append((0, 0))
    visited[0][0] = True

    while q:
        cx, cy = q.popleft()
        move = m[cx][cy]
        nx, ny = cx + move, cy + move

        if 0 <= nx < N and not visited[nx][cy]:
            visited[nx][cy] = True
            q.append((nx, cy))
        if 0 <= ny < N and not visited[cx][ny]:
            visited[cx][ny] = True
            q.append((cx, ny))

    if visited[N - 1][N - 1]:
        return True
    else:
        return False

N = int(input())
m = [[] * N for _ in range(N)]
visited = [[False] * N for _ in range(N)]

for i in range(N):
    m[i] = list(map(int, input().split()))

if bfs():
    print("HaruHaru")
else:
    print("Hing")